### DISCUSSION TITLE
3D models guidelines

### DISCUSSION THREAD

**Discussion by 12gabriel3** [EveronLife/EveronLife#99] [Ideas]
IMO we need more performance metrics for 3d models and textures other than disk space. I suggest these metrics (for small items), based on my experience:
Less than 1k triangles for LOD0
HAVE LODS! This is a must, at least 2 extra LODs other than LOD0
The last lod must not go beyond 100 triangles, the least the better
Texture 256 max. You should use techniques to reduce the need of bigger textures such as overlapping UV islands and tiling. My apple looks just fine with a 128 texture
Im not sure about the technical details, but one 256 texture has the same amount of pixels as 4 128 textures, so prefer to use up to 3 materials with 128 textures before you consider a 256 texture
Use mips (its on by default so no extra work here)

These are the most common optimisations I came across when making my models and I think it should be the bare minimum requirements for a model to be accepted. We dont want our mod to be a lag fest. 

I suggest someone more experienced than I am to make some guidelines for the models, but the things i mentioned are very important and lacking on many models

**Comment by TheOriginalCrazyDuck** [comment]
I don't think hard limits on the number of triangles or vertices would make sense as a tomato made from 1k triangles could be incredibly detailed but a more complex object like a building (like the prison project) would have to be far less detailed to fit within the guidelines so perhaps a limit that scales would be better, for reference, even the Land Rover in Arma 2 OA used 15k Triangles for LOD0

**Reply by LuxPoseidon** [reply]
generally speaking triangles aren't even the biggest issue for performance, really just make sure textures are packed and for like a building with a bunch of textures should be packed into 1-2 textures as this will reduce calls that are way more of an issue than triangles.

**Comment by TL6868** [comment]
I 100% agree I feel that the best way to figure out these numbers is to look at the base assets so that we can then set standards for different types of objects like trees, buildings, cars, and weapons. The base game is optimized well enough so that could help to give the framework similar performance to the base game

**Comment by Arkensor** [comment]
I think 1k triangles must not be a hard limit but it sure helps people to have an orientation. File sizes are our primary concern. The engine can handle "detailed" objects. As long as someone does not add a bucket that has 250k triangles 😄 

Some thoughts I want to add to the guidelines (wording not final):
- Triple-check that the base game or this repository does not already have an asset that could be re-purposed with new material/texture.
- Only use `.fbx` and `.tiff`. Those files must be included alongside the source as well then.
- Textures and materials need to be suffixed according to the slot they are made for e.g. `Name_BCR.edds`, `Name_NMO.edds`. The suffix serves a technical purpose for the game to select compression algorithms. Check out the vanilla game assets for reference.
	- [Biki on Textures and compression in general](https://community.bistudio.com/wiki/Arma_Reforger:Textures )
	- The [Arma Reforger Discord](https://discord.gg/arma) contains additional (often pinned) information. For textures (and materials) check out `#enfusion_texture_markers`
		- [Texture compression explaination](https://discordapp.com/channels/105462288051380224/976231628785475624/981871081046569000)
		- [Devs use TIFF with 8Bits and LZW](https://discordapp.com/channels/105462288051380224/976231628785475624/977200057357074432)
		- [Substance painter export remarks](https://discordapp.com/channels/105462288051380224/976231628785475624/977154368526753812)
- PascalCase convention for all files **and folders**. Only use `_` for grouping objects e.g. `Radio_ANPRC68` vs `Radio_R107M`
	- **BAD:** `Assets\apple_tree\apple.xob` -> **GOOD:** `Assets\AppleTree\Apple.xob`. 
	- **BAD:** `Assets\AppleTree\data\` -> **GOOD:** `Assets\AppleTree\Data\`. This one is particularly important because the *"very nice to use and bug-free /s"* workbench automatically creates a `data` folder instead of `Data` that is used in all base game directories. This is [reported as a bug](https://feedback.bistudio.com/T165764). In the meantime **Create the Data folder manually before importing the .fbx**. Changing it later messes up all the meta files and other paths where it gets automatically referenced.
- Please do not commit working names such as `thing_01.xob` or `stuff_test_v2.xob`. If you are unsure what to call something or where to place it, reach out to maintainers, please! You are unable to effectively rename them later without redoing the whole import, prefab creation, and world edits again!

**Comment by Kanf** [comment]
I agree with Arkensor, we don't need hard limits on polycount, but suggestions are good idea. For example, rifles in Reforger range from ~18k tris to ~40k tris for LOD0 and it runs just fine. Reasonably high polycount is especially important for LOD0 for items held in hand, as they take up big part of the screen. But I have to completely disagree with you @12gabriel3 on using 3 materials instead of 1 higher resolution, usually draw calls are much worse for perfomance than loading few more megabytes of texture, or at least that's how (from what I know) it works in Unity and Unreal. More important will be using tiling materials for large assets like buildings.

**Reply by 12gabriel3** [reply]
I haven't tested it. My reasoning was that one 256 texture is the size of 4 128 textures. Thats why I think we should have someone experienced to make the guidelines. 
About poly count, we should not have a highly detailed weapon as a reference for an apple, for example. It all depends on the mesh topology, if it is well made. You souldn't use 40k tris on a pencil, for example, because it is a much more simple object than a gun.

**Reply by Kanf** [reply]
It's all good. I agree we should not mindlessly use polygons on props that don't need them, I just meant we don't need to worry too much about using more polygons where they are needed. We should sit down together at some point and make guidelines for artists.

**Reply by Arkensor** [reply]
> We should sit down together at some point and make guidelines for artists.

This is what I wanted to get going with this discussion yes. But feel free to sit down in voice, as long as you write down your final result of it here please :)
